### PR TITLE
[RISCV][SDAG] Prefer ShortForwardBranch to lower sdiv by pow2

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch lowers `sdiv x, +/-2**k` to `add + select + shift` when the short forward branch optimization is enabled. The latter inst seq performs faster than the seq generated by target-independent DAGCombiner. This algorithm is described in ***Hacker's Delight***.

This patch also removes duplicate logic in the X86 and AArch64 backend. But we cannot do this for the PowerPC backend since it generates a special instruction `addze`.

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 1]
```diff
@@ -16102,33 +16102,7 @@ AArch64TargetLowering::BuildSDIVPow2(SDNode *N, const APInt &Divisor,
...
 
-  SDLoc DL(N);
-  SDValue N0 = N->getOperand(0);
-  unsigned Lg2 = Divisor.countr_zero();
-  SDValue Zero = DAG.getConstant(0, DL, VT);
-  SDValue Pow2MinusOne = DAG.getConstant((1ULL << Lg2) - 1, DL, VT);
-
-  // Add (N0 < 0) ? Pow2 - 1 : 0;
-  SDValue CCVal;
-  SDValue Cmp = getAArch64Cmp(N0, Zero, ISD::SETLT, CCVal, DAG, DL);
```
Comment (topperc) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:16113]: The AArch64 code was using target specific nodes not generic nodes. This changes needs to be reviewed by AArch64 developers.

Comment (dtcxzyw) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:16113]: Ping for review from AArch64 developers.

Comment (dtcxzyw) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:16113]: Ping for review from AArch64 developers.

File: llvm/test/CodeGen/RISCV/sdiv-pow2-cmov.ll [hunk 2]
```diff
@@ -0,0 +1,386 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
```
Comment (jrtc27) [inline llvm/test/CodeGen/RISCV/sdiv-pow2-cmov.ll:1]: Please precommit tests

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 1 above)
Comment (davemgreen) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:16113]: Sorry for the delay, I didn't see this. I think it sounds fine so long as it is usually expected to optimize to the same thing.

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 3]
```diff
@@ -6009,6 +6009,48 @@ TargetLowering::BuildSREMPow2(SDNode *N, const APInt &Divisor,
...
+  EVT VT = N->getValueType(0);
+
+  SDLoc DL(N);
+  SDValue N0 = N->getOperand(0);
+  SDValue Zero = DAG.getConstant(0, DL, VT);
+  APInt Lg2Mask = APInt::getLowBitsSet(VT.getSizeInBits(), Lg2);
+  SDValue Pow2MinusOne = DAG.getConstant(Lg2Mask, DL, VT);
+
+  // If N0 is negative, we need to add (Pow2 - 1) to it before shifting right.
+  SDValue Cmp = DAG.getSetCC(DL, MVT::i8, N0, Zero, ISD::SETLT);
```
Comment (davemgreen) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:6033]: I may have spoken to soon. I don't think this should be an i8 for AArch64. It needs to be the getSetCCResultType(VT)?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 4]
```diff
@@ -19304,6 +19304,26 @@ unsigned RISCVTargetLowering::getCustomCtpopCost(EVT VT,
...
+    return SDValue(N, 0); // Lower SDIV as SDIV
+
+  // Only perform this transform if short forward branch opt is supported.
+  if (!Subtarget.hasShortForwardBranchOpt())
+    return SDValue();
+  EVT VT = N->getValueType(0);
+  if (!(VT == MVT::i32 || (VT == MVT::i64 && Subtarget.is64Bit())))
+    return SDValue();
+  unsigned Lg2 = Divisor.countr_zero();
+  // ensure 2**k-1 < 2048
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:19322]: Why not compare the divisor to 2048? Why do we need to count the trailing zeros?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 5]
```diff
@@ -19304,6 +19304,26 @@ unsigned RISCVTargetLowering::getCustomCtpopCost(EVT VT,
...
+  if (Lg2 >= 11)
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:19323]: Probably better to check `Divisor > 2048 || Division < -2048`? The Lg2 check made sense when Lg2 was needed later, but that's in TargetLowering::buildSDIVPow2WithCMov now.

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 6]
```diff
@@ -22714,32 +22714,7 @@ X86TargetLowering::BuildSDIVPow2(SDNode *N, const APInt &Divisor,
   if (Lg2 == 1)
```
Comment (topperc) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:22714]: Just check the `Divisor == 2 || Division == -2`. The Lg2 made mode sense when we already needed Lg2 for the rest of the code, but that's in the TargetLowering::buildSDIVPow2WithCMov now.

**Comment by dtcxzyw** [comment]
Ping @topperc @davemgreen

**Review by topperc — APPROVED** [review]
LGTM

**Comment by dtcxzyw** [comment]
Any more comments? @davemgreen

**Comment by davemgreen** [comment]
Yeah I think I'm happy. So long as nothing is getting worse then this LGTM. Thanks

**Pull request merged**